builder/transaction_builder.go

```diff
@@ -676,62 +690,127 @@ func (b *TransactionBuilder) Build(signer iotago.AddressSigner) (*iotago.SignedT
+		owner := b.inputOwner[inputRef.(*iotago.UTXOInput).OutputID()]
+
+		chainAddr, isChainAddress := owner.(iotago.ChainAddress)
+		if isChainAddress {
+			// the inputs's owning chain address must have been unlocked already
+			unlockedAtIndex, isUnlocked := unlockedSet.isChainUnlocked(chainAddr)
+			if !isUnlocked {
+				return nil, ierrors.Errorf("input %d's owning chain is not unlocked, chainID %s, type %s", inputIndex, owner.Bech32(b.api.ProtocolParameters().Bech32HRP()), owner.Type())
 			}
 
-			// produce signature
+			// add a referential unlock to the former unlock position
+			unlockedSet.addReferentialUnlock(owner, unlockedAtIndex)
+
+			// always mark the chain as unlocked in case the output is a chain output
+			unlockedSet.addChainAsUnlocked(inputs[inputIndex], inputIndex)
+
+			// skip the rest of the input processing because we don't need to sign chain inputs
+			continue
+		}
+
+		if _, isDirectUnlockable := owner.(iotago.DirectUnlockableAddress); !isDirectUnlockable {
+			// we only support directly unlockable addresses in the transaction builder for now
+			return nil, ierrors.Errorf("input %d's owning address is not directly unlockable, address %s, type %s", inputIndex, owner.Bech32(b.api.ProtocolParameters().Bech32HRP()), owner.Type())
+		}
+
+		// get the signer UID for the directly unlockable address
+		signerUID, err := b.signer.SignerUIDForAddress(owner)
+		if err != nil {
+			return nil, ierrors.Wrapf(err, "failed to get signer UID for address %s", owner.Bech32(b.api.ProtocolParameters().Bech32HRP()))
+		}
+
+		unlockedAtIndex, alreadyUnlocked := unlockedSet.signerUIDs[signerUID]
+		if !alreadyUnlocked {
+			var err error
 			var signature iotago.Signature
-			signature, err = signer.Sign(addr, txEssenceData)
+			if signEssence {
+				// sign the tx essence data
+				signature, err = b.signer.Sign(owner, txEssenceData)
+			} else {
+				// sign with empty signature.
+				// this is used for example to calculate the workscore of the transaction before the actual signing.
+				signature, err = b.signer.EmptySignatureForAddress(owner)
+			}
 			if err != nil {
 				return nil, ierrors.Wrapf(err, "failed to sign tx transaction: %s", txEssenceData)
 			}
 
-			unlocks = append(unlocks, &iotago.SignatureUnlock{Signature: signature})
-			addChainAsUnlocked(inputs[i], i, unlockPos)
-			unlockPos[addrKey] = i
+			// add the new signature to the unlocks
+			unlockedSet.addUnlock(&iotago.SignatureUnlock{Signature: signature})
 
-			continue
+			// remember the unlock index for the new signer UID
+			unlockedSet.addSignerUID(signerUID, inputIndex)
+		} else {
+			// add a referential unlock to the former unlock position
+			unlockedSet.addReferentialUnlock(owner, unlockedAtIndex)
 		}
 
-		unlocks = addReferentialUnlock(addr, unlocks, pos)
-		addChainAsUnlocked(inputs[i], i, unlockPos)
+		// always mark the chain as unlocked in case the output is a chain output
+		unlockedSet.addChainAsUnlocked(inputs[inputIndex], inputIndex)
```

Comment:
Can we have chain outputs unlocked by chain address? I recall something about this not being allowed...